cpp/src/utility/quadtree_thrust.cuh

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <thrust/pair.h>
+#include <thrust/tuple.h>
+#include <thrust/functional.h>
+#include "z_order.cuh"
+#include "bbox_thrust.cuh"
+namespace 
+{
+
+template<typename T>
+struct xytoz 
+{
+
+    SBBox<T> bbox;
+    uint8_t lev;
+    double scale;
+
+    xytoz(SBBox<T> _bbox,uint8_t _lev,double _scale): bbox(_bbox),lev(_lev),scale(_scale) {}
+
+    __device__
+    uint32_t operator()(thrust::tuple<double,double> loc )
+    {
+        double x=thrust::get<0>(loc);
+        double y=thrust::get<1>(loc);
+        if(x<thrust::get<0>(bbox.first)||x>thrust::get<0>(bbox.second)||y<thrust::get<1>(bbox.first)||y>thrust::get<1>(bbox.second))
+            return (1<<(2*lev)-1);
```

Comment:
@zhangjianting I'm getting an ambiguity warning from clang on this line.

Should this be `1 << ((2*lev) - 1)` or `(1 << (2*lev)) - 1`?

Reply:
I assumed << has precedence than -. What I wanted is (1 << (2*lev)) - 1.  Interesting that clang considers << has the same precedence as -

Reply:
Here's the actual clang warning:
```
Operator '<<' has lower precedence than '-'; '-' will be evaluated first (fix available)clang(-Wshift-op-parentheses)
```

It's recommended fix is `1 << ((2 * lev) - 1)`, so I'm glad I asked for clarification!

Reply:
Just learnt that subtraction is one level higher than shift. I need to be more careful on this.  
In practice, since very few points and polygons are distributed at the the last quadrant (among (2^(max_level-1)) quadrants) in the data I have tested, the corner case did not happen.